src/vm/analysis/type_checker/natives/sequences.rs

```diff
@@ -126,33 +133,40 @@ pub fn check_special_concat(checker: &mut TypeChecker, args: &[SymbolicExpressio
 
     analysis_typecheck_cost(checker, &lhs_type, &rhs_type)?;
     
-    match lhs_type {
-        TypeSignature::ListType(lhs_list) => {
-            if let TypeSignature::ListType(rhs_list) = rhs_type {
-                let (lhs_entry_type, lhs_max_len) = lhs_list.destruct();
-                let (rhs_entry_type, rhs_max_len) = rhs_list.destruct();
-
-                let list_entry_type = TypeSignature::least_supertype(&lhs_entry_type, &rhs_entry_type)?;
-                let new_len = lhs_max_len.checked_add(rhs_max_len)
-                    .ok_or(CheckErrors::MaxLengthOverflow)?;
-                let return_type = TypeSignature::list_of(list_entry_type, new_len)?;
-                return Ok(return_type);
-            } else {
-                return Err(CheckErrors::TypeError(rhs_type.clone(), TypeSignature::ListType(lhs_list)).into());
-            }
-        },
-        TypeSignature::BufferType(lhs_buff_len) => {
-            if let TypeSignature::BufferType(rhs_buff_len) = rhs_type {
-                let size: u32 = u32::from(lhs_buff_len).checked_add(u32::from(rhs_buff_len))
-                    .ok_or(CheckErrors::MaxLengthOverflow)?;
-                let return_type = TypeSignature::BufferType(size.try_into()?);
-                return Ok(return_type);
-            } else {
-                return Err(CheckErrors::TypeError(rhs_type.clone(), TypeSignature::max_buffer()).into());
+    let res = match (&lhs_type, &rhs_type) {
```

Comment:
Can this be an ownerhsip consuming match rather than a reference match? It allows us to get away with a little bit fewer clones.

Reply:
Happy to start a clone war here, but I can't spot the occurrences you have in mind, can you provide a pointer? thank you!

Reply:
Oops -- I was mistaken, I thought that the `get_list_item_type()` function returned a clone of the entry type, but it's a reference.